File: translate/google_translate.go
package translate

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

var (
	translationCache = struct {
		sync.RWMutex
		m map[string]string
	}{m: make(map[string]string)}
	lastRequest time.Time
	requestLock sync.Mutex
)

func encodeURI(s string) string {
	return url.QueryEscape(s)
}

func TranslateText(source, sourceLang, targetLang string) (string, error) {
	// Проверяем кэш
	cacheKey := sourceLang + targetLang + source
	translationCache.RLock()
	if cached, exists := translationCache.m[cacheKey]; exists {
		translationCache.RUnlock()
		return cached, nil
	}
	translationCache.RUnlock()

	// Ограничение запросов (1 запрос в секунду)
	requestLock.Lock()
	elapsed := time.Since(lastRequest)
	if elapsed < time.Second {
		time.Sleep(time.Second - elapsed)
	}
	lastRequest = time.Now()
	requestLock.Unlock()

	encodedSource := encodeURI(source)
	url := "https://translate.googleapis.com/translate_a/single?client=gtx&sl=" +
		sourceLang + "&tl=" + targetLang + "&dt=t&q=" + encodedSource

	resp, err := http.Get(url)
	if err != nil {
		return "", fmt.Errorf("error connecting to translate API: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("translate API returned status: %s", resp.Status)
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("error reading response: %v", err)
	}

	if strings.Contains(string(body), `<title>Error 400 (Bad Request)`) {
		return "", errors.New("bad request to translate API")
	}

	var result []interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return "", fmt.Errorf("error parsing response: %v", err)
	}

	if len(result) == 0 {
		return "", errors.New("empty response from translate API")
	}

	var translatedText strings.Builder
	if slice, ok := result[0].([]interface{}); ok {
		for _, item := range slice {
			if pair, ok := item.([]interface{}); ok && len(pair) > 0 {
				if text, ok := pair[0].(string); ok {
					translatedText.WriteString(text)
				}
			}
		}
	}

	if translatedText.Len() == 0 {
		return "", errors.New("no translation found in response")
	}

	// Сохраняем в кэш
	translationCache.Lock()
	translationCache.m[cacheKey] = translatedText.String()
	translationCache.Unlock()

	return translatedText.String(), nil
}

File: translate/google_translate.py
import json
import time
import urllib.parse
import urllib.request
from threading import Lock

_translation_cache = {}
_cache_lock = Lock()
_last_request_time = 0
_request_lock = Lock()


def translate_text(source, source_lang, target_lang):
    """Переводит текст с помощью Google Translate API"""
    global _translation_cache, _last_request_time
    
    # Проверяем кэш
    cache_key = f"{source_lang}{target_lang}{source}"
    with _cache_lock:
        if cache_key in _translation_cache:
            return _translation_cache[cache_key]
    
    # Ограничение запросов (1 запрос в секунду)
    with _request_lock:
        elapsed = time.time() - _last_request_time
        if elapsed < 1.0:
            time.sleep(1.0 - elapsed)
        _last_request_time = time.time()
    
    # Кодируем текст для URL
    encoded_source = urllib.parse.quote(source)
    url = (f"https://translate.googleapis.com/translate_a/single?"
           f"client=gtx&sl={source_lang}&tl={target_lang}&dt=t&q={encoded_source}")
    
    try:
        with urllib.request.urlopen(url) as response:
            if response.status != 200:
                raise Exception(f"Translate API returned status: {response.status}")
            
            body = response.read().decode('utf-8')
            
            if '<title>Error 400 (Bad Request)' in body:
                raise Exception("Bad request to translate API")
            
            result = json.loads(body)
            
            if not result or len(result) == 0:
                raise Exception("Empty response from translate API")
            
            translated_text = ""
            if isinstance(result[0], list):
                for item in result[0]:
                    if isinstance(item, list) and len(item) > 0:
                        if isinstance(item[0], str):
                            translated_text += item[0]
            
            if not translated_text:
                raise Exception("No translation found in response")
            
            # Сохраняем в кэш
            with _cache_lock:
                _translation_cache[cache_key] = translated_text
            
            return translated_text
    
    except Exception as e:
        raise Exception(f"Error translating text: {e}")

